File: archive/src_java_applet/CTruthTable.java

/**
 * Title:        TruthTable <p>
 * Description:  <p>
 * Copyright:    Copyright (c) Joe Triscari<p>
 * Company:      UWM<p>
 * @author Joe Triscari
 * @version 1.0
 */
package truthtable;

public class CTruthTable {
  private CExpression m_Expression;
  private boolean m_bFullTable;
  private char m_cTrueSymbol;
  private char m_cFalseSymbol;

  public CTruthTable(CExpression e)
  {
    m_Expression = e;
    m_bFullTable = false;
    m_cTrueSymbol = 'T';
    m_cFalseSymbol = 'F';
  }

  public String printTable(boolean bStyle)
  {
    long lCombos=0;
    LinkList lstOperands=null;
    LinkList lstExpr=null;
    String strPrint="";
    int len = 0;

    m_bFullTable = bStyle;
    lCombos = m_Expression.GetCombonationCount();
    lstOperands = m_Expression.GetOperands();

    // PRINT THE OPERANDS.
    lstOperands.moveFirst();
    while(!lstOperands.EOL())
    {
      Opdata d = (Opdata)lstOperands.getData();
      strPrint += String.valueOf(d.cOperand);
      lstOperands.moveNext();
    }

    if(m_bFullTable == false)  // PRINT THE EXPRESSION.
    {
       strPrint += "| " + m_Expression.GetExpression() + "\n";

       // PRINT THE HORIZONTAL BORDER.
       for(int i=0;  i < m_Expression.GetVariableCount() +
                     m_Expression.GetExpression().length() + 2; i++)
         strPrint += (i == m_Expression.GetVariableCount() ? "|" : "-");

       strPrint += "\n";
    }
    else  // PRINT ALL THE SUB EXPRESSIONS.
    {
      lstExpr = m_Expression.GetSubExpressions();
      int iCount=0;
      lstExpr.moveFirst();
      while(!lstExpr.EOL())
      {
         strPrint += (iCount==0 ? "| " : " | ") +
                     (String)lstExpr.getData();
         ++iCount;
         lstExpr.moveNext();
      }
      strPrint += "\n";

      for(int i=0; i<m_Expression.GetVariableCount(); i++)
        strPrint += "-";

      strPrint += "|";
      lstExpr.moveFirst();
      while(!lstExpr.EOL())
      {
          len = ((String)lstExpr.getData()).length();
          for(int i=0; i<len+2; i++)
            strPrint += "-";
          if(lstExpr.getNext() != null) strPrint += "|";
          lstExpr.moveNext();
      }
      strPrint += "\n";
   }

    //for(long i=0; i<lCombos; i++)
   for(long i=lCombos-1; i >= 0; i--)
   {
      for(int k=0; k<m_Expression.GetVariableCount(); k++)
        if( (lCombos & (i << k+1)) != 0)
          strPrint += String.valueOf(m_cTrueSymbol);
        else
          strPrint += String.valueOf(m_cFalseSymbol);

      if(m_bFullTable == false)
      {
         strPrint += "| ";
         if(m_Expression.GetTruthValue(i)==1)
           strPrint += String.valueOf(m_cTrueSymbol);
         else
           strPrint += String.valueOf(m_cFalseSymbol);
         strPrint += "\n";
      }
      else
      {
         LinkList lstVal = m_Expression.GetTruthValues(i);
         lstExpr.moveFirst();
         strPrint += "| ";
         lstVal.moveFirst();
         while(!lstVal.EOL())
         {
            Integer iTmp = (Integer)lstVal.getData();
            if(iTmp.intValue() == 1)
              strPrint += String.valueOf(m_cTrueSymbol);
            else
              strPrint += String.valueOf(m_cFalseSymbol);

            if(lstVal.getNext() != null)
            {
               len = ((String)lstExpr.getData()).length();
               for(int j=0; j<len; j++)
                 strPrint += " ";
               strPrint += "| ";
            }
            lstVal.moveNext();
            lstExpr.moveNext();
         }
         strPrint += "\n";
      }

   }

   return strPrint;
}

} // End of class

File: archive/src_cpp/CTruthTable.cpp
/*********************************************************/
/****  CTruthTable.cc                                 ****/
/****                                                 ****/
/****                                                 ****/
/****  Author:  Joseph N. Triscari                    ****/
/****  Email:   triscaj@yahoo.com                     ****/
/****  Created: 10/5/2000                             ****/
/****                                                 ****/
/****  Description:                                   ****/
/****    This file contains the class implementation  ****/
/****    for CTruthTable.                             ****/
/****                                                 ****/
/****                                                 ****/
/****  An unpublished work. All rights reserved.      ****/
/****                                                 ****/
/****                                                 ****/
/****                                                 ****/
/*********************************************************/

#include "CTruthTable.h"

void CTruthTable::printTable(ostream* out, TableStyle style, char cTrueSymbol, char cFalseSymbol)
{
   long lCombos=0;
   register OperandList * pOperands=NULL;   
   ExpressionList * pExpr=NULL;

   if(style != Default) m_eTableStyle = style;

   lCombos = m_Expression.GetCombonationCount();
   pOperands = m_Expression.GetOperands();

   // PRINT THE OPERANDS.
   while(pOperands) { (*out) << pOperands->cOperand; pOperands = pOperands->pNext; }

   if(m_eTableStyle == Normal)  // PRINT THE EXPRESSION.
   {
      (*out) << "| " << m_Expression.GetExpression() << endl;

      // PRINT THE HORIZONTAL BORDER.
      for(int i=0; i<m_Expression.GetVariableCount() + m_Expression.GetExpression().length() + 2; i++)
         (*out) << (i == m_Expression.GetVariableCount() ? '|' : '-');

      (*out) << endl;
   }
   else  // PRINT ALL THE SUB EXPRESSIONS.
   {
      pExpr = m_Expression.GetSubExpressions();
      int iCount=0;
      while(pExpr)
      {
         (*out) << (iCount==0 ? "| " : " | ") << pExpr->strExpression;
         ++iCount;
         pExpr = pExpr->pNext;
      }
      (*out) << endl;

      for(int i=0; i<m_Expression.GetVariableCount(); i++)
        (*out) << '-';

      (*out) << '|';
       pExpr = m_Expression.GetSubExpressions();
       while(pExpr)
       {
          for(int i=0; i<pExpr->strExpression.length()+2; i++) (*out) << '-';
          if(pExpr->pNext) (*out) << '|';
          pExpr = pExpr->pNext;
       }
       (*out) << endl;
   }

   //for(register long i=0; i<lCombos; i++)
   for(register long i=lCombos-1; i>=0; i--)
   {
      for(register int k=0; k<m_Expression.GetVariableCount(); k++)
        if( (lCombos & (i << k+1)) )
          (*out) << cTrueSymbol; 
        else
          (*out) << cFalseSymbol;

      if(m_eTableStyle==Normal)
         (*out) << "| " << (m_Expression.GetTruthValue(i) ? cTrueSymbol : cFalseSymbol) << endl;
      else
      {
         ValueList * pVal=m_Expression.GetTruthValues(i);
         pExpr = m_Expression.GetSubExpressions();
         (*out) << "| ";
         while(pVal)
         {
            (*out) << (pVal->iValue ? cTrueSymbol : cFalseSymbol);
            if(pVal->pNext)
            {
               for(int j=0; j<pExpr->strExpression.length(); j++) (*out) << ' ';
               (*out) << "| ";
            }
            pVal = pVal->pNext;
            pExpr = pExpr->pNext;
         }
         (*out) << endl;
      }

   }

}
